app/src/main/java/com/cmput301w23t40/capturetheqr/ScoreboardActivity.java

```diff
@@ -39,125 +57,138 @@ public class ScoreboardActivity extends AppCompatActivity implements AdapterView
-                        } else if (itemSelected.equals("Highest number of QRs scanned")) {
-                            Collections.sort(allPlayers, (o1, o2) -> Integer.compare(o2.getNumberOfCodes(), o1.getNumberOfCodes()));
-                        } else if (itemSelected.equals("Highest sum of QR scores")) {
-                            Collections.sort(allPlayers, (o1, o2) -> Integer.compare(o2.getScoreSum(), o1.getScoreSum()));
-                        }
-                        for(int i = 1; i <= allPlayers.size(); ++i){
-                            Player tempPlayer;
-                            tempPlayer = allPlayers.get(i-1);
-                            tempPlayer.setRank(i);
-                            if (tempPlayer.getDeviceID().equals(deviceID)){
-                                my_player = tempPlayer;
+                        if(!itemSelected.equals("Nearby Highest Scoring QR code")){
+                            if (itemSelected.equals("Highest QR score")) {
+                                Collections.sort(allPlayers, (o1, o2) -> Integer.compare(o2.getHighScore(), o1.getHighScore()));
+                                playerAdapter = new ScoreboardList(getApplicationContext(), allPlayers, SortBy.HIGHEST_SCORE);
+                            } else if (itemSelected.equals("Highest number of QRs scanned")) {
+                                Collections.sort(allPlayers, (o1, o2) -> Integer.compare(o2.getNumberOfCodes(), o1.getNumberOfCodes()));
+                                playerAdapter = new ScoreboardList(getApplicationContext(), allPlayers, SortBy.NUMBER_OF_CODES);
+                            } else if (itemSelected.equals("Highest sum of QR scores")) {
+                                Collections.sort(allPlayers, (o1, o2) -> Integer.compare(o2.getScoreSum(), o1.getScoreSum()));
+                                playerAdapter = new ScoreboardList(getApplicationContext(), allPlayers, SortBy.SCORE_SUM);
                             }
+                            boolean foundMyPlayer = false;
+                            for(int i = 0; i < allPlayers.size(); ++i){
+                                Player tempPlayer = allPlayers.get(i);
+                                tempPlayer.setRank(i+1);
+                                if(!foundMyPlayer && tempPlayer.getDeviceID().equals(deviceID)){
+                                    my_player = tempPlayer;
+                                    foundMyPlayer = true;
+                                }
+                            }
+                            playerAdapter.notifyDataSetChanged();
+                            listView.setAdapter(playerAdapter);
+                            setMyRankText(my_player.getRank(), itemSelected);
+                        } else {
+                            updateLocation();
+                            locationHelper.updateLocation(new PlayerLocation.CallbackLocation() {
```

Comment:
locationHelper will be null here if the user does not grant location permissions, as the only places it is initialized is inside blocks where we have those permissions.

Reply:
done